client/client.go

```diff
@@ -247,6 +247,14 @@ func (cli *Client) tlsConfig() *tls.Config {
 
 func defaultHTTPClient(hostURL *url.URL) (*http.Client, error) {
 	transport := &http.Transport{}
+	// Necessary to prevent long-lived processes using the
+	// client from leaking connections due to idle connections
+	// not being released.
+	// TODO: see if we can also address this from the server side,
+	// or in go-connections.
+	// see: https://github.com/moby/moby/issues/45539
+	transport.MaxIdleConns = 6
+	transport.IdleConnTimeout = 30 * time.Second
```

Comment:
These values seem fairly low. Maybe we could keep them the same as defaults in `http.DefaultTransport`?

```
var DefaultTransport RoundTripper = &Transport{
    ...
	MaxIdleConns:          100,
	IdleConnTimeout:       90 * time.Second,
...
```
https://cs.opensource.google/go/go/+/refs/tags/go1.23.2:src/net/http/transport.go;l=51-52

Reply:
I feel like 100 idle connections is a lot, especially when those are all open FDs, but I'm open to changing it. Thoughts? @thaJeztah @corhere @crazy-max ?

Reply:
Maybe we don't need as much as 100, but 6 feels a bit too conservative IMO 😅 These are still subject to the `IdleConnTimeout` so I think we can afford to go a little bit higher than that.

Reply:
I guess for regular use in the CLI it wouldn't make a huge difference, but possibly more so if the client is used in a long-lived process. Looking at the defaults, `MaxIdleConns` is unlimited; https://github.com/golang/go/blob/2ef8e41f9543478a51a0147a735e4415737de09f/src/net/http/transport.go#L205-L207

```go
// MaxIdleConns controls the maximum number of idle (keep-alive)
// connections across all hosts. Zero means no limit.
MaxIdleConns int
```

But `MaxIdleConnsPerHost` defaults to 2 (`DefaultMaxIdleConnsPerHost`)

https://github.com/golang/go/blob/2ef8e41f9543478a51a0147a735e4415737de09f/src/net/http/transport.go#L209-L212

```go
// MaxIdleConnsPerHost, if non-zero, controls the maximum idle
// (keep-alive) connections to keep per-host. If zero,
// DefaultMaxIdleConnsPerHost is used.
MaxIdleConnsPerHost int
```

https://github.com/golang/go/blob/2ef8e41f9543478a51a0147a735e4415737de09f/src/net/http/transport.go#L58-L60

```go
// DefaultMaxIdleConnsPerHost is the default value of [Transport]'s
// MaxIdleConnsPerHost.
const DefaultMaxIdleConnsPerHost = 2
```

Which would mean that with the current settings, the client would have to have idle connections for 5 different hosts before this limit kicks in. Raising to 100 _without_ raising `DefaultMaxIdleConnsPerHost` means that the client would need to have iddle connections for 50 different hosts before it taking effect. That may be very unlikely (as hosts here would be "Engine APIs" or perhaps "registry to authenticate").

Also did a check for prior art in our codebase; here's some defaults that are set by SDKs we use;


```bash
git grep -E '(MaxIdleConns|IdleConnTimeout)[^!]*[:=]'
plugin/registry.go:                                                                     IdleConnTimeout:     30 * time.Second,
vendor/cloud.google.com/go/compute/metadata/CHANGES.md:* **compute/metadata:** Set IdleConnTimeout for http.Client ([#7084](https://github.com/googleapis/google-cloud-go/issues/7084)) ([766516a](https://github.com/googleapis/google-cloud-go/commit/766516aaf3816bfb3159efeea65aa3d1d205a3e2)), refs [#5430](https://github.com/googleapis/google-cloud-go/issues/5430)
vendor/cloud.google.com/go/compute/metadata/metadata.go:                                IdleConnTimeout: 60 * time.Second,
vendor/github.com/aws/aws-sdk-go-v2/aws/transport/http/client.go:                       DefaultHTTPTransportMaxIdleConns        = 100
vendor/github.com/aws/aws-sdk-go-v2/aws/transport/http/client.go:                       DefaultHTTPTransportMaxIdleConnsPerHost = 10
vendor/github.com/aws/aws-sdk-go-v2/aws/transport/http/client.go:                       DefaultHTTPTransportIdleConnTimeout       = 90 * time.Second
vendor/github.com/aws/aws-sdk-go-v2/aws/transport/http/client.go:                       MaxIdleConns:          DefaultHTTPTransportMaxIdleConns,
vendor/github.com/aws/aws-sdk-go-v2/aws/transport/http/client.go:                       MaxIdleConnsPerHost:   DefaultHTTPTransportMaxIdleConnsPerHost,
vendor/github.com/aws/aws-sdk-go-v2/aws/transport/http/client.go:                       IdleConnTimeout:       DefaultHTTPTransportIdleConnTimeout,
vendor/github.com/moby/buildkit/util/resolver/resolver.go:                              MaxIdleConns:          30,
vendor/github.com/moby/buildkit/util/resolver/resolver.go:                              IdleConnTimeout:       120 * time.Second,
vendor/github.com/moby/buildkit/util/resolver/resolver.go:                              MaxIdleConnsPerHost:   4,
vendor/go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp/client.go:     MaxIdleConns:          100,
vendor/go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp/client.go:     IdleConnTimeout:       90 * time.Second,
vendor/go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp/client.go:       MaxIdleConns:          100,
vendor/go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp/client.go:       IdleConnTimeout:       90 * time.Second,
vendor/google.golang.org/api/internal/creds.go:                                         MaxIdleConns:          100,
vendor/google.golang.org/api/internal/creds.go:                                         MaxIdleConnsPerHost:   100,
vendor/google.golang.org/api/internal/creds.go:                                         IdleConnTimeout:       90 * time.Second,
vendor/google.golang.org/api/transport/http/dial.go:                                    trans.MaxIdleConnsPerHost = 100
vendor/google.golang.org/api/transport/http/dial.go:                                    MaxIdleConns:          100,
vendor/google.golang.org/api/transport/http/dial.go:                                    MaxIdleConnsPerHost:   100,
vendor/google.golang.org/api/transport/http/dial.go:                                    IdleConnTimeout:       90 * time.Second,
```

Reply:
Hmm right, I missed that `MaxIdleConnsPerHost` defaults to 2.